std/algorithm/comparison.d

```diff
@@ -729,43 +736,109 @@ template equal(alias pred = "a == b")
     See_Also:
         $(WEB sgi.com/tech/stl/_equal.html, STL's _equal)
     +/
-    bool equal(Range1, Range2)(Range1 r1, Range2 r2)
-    if (isInputRange!Range1 && isInputRange!Range2 && is(typeof(binaryFun!pred(r1.front, r2.front))))
+    bool equal(R, Ss...)(R r, Ss ss)
+    if (ss.length >= 1 &&
+        isInputRange!R &&
+        allSatisfy!(isInputRange, Ss))
     {
-        //Start by detecting default pred and compatible dynamicarray.
-        static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isArray!Range1 && isArray!Range2 && is(typeof(r1 == r2)))
+        enum hasLengthComparableToR(T) = haveEquableLengths!(R, T);
+        enum isEquableToR(T) = areEquable!(R, T);
+
+        // start by detecting default pred and compatible dynamicarray
+        static if (is(typeof(pred) == string) &&
+                   pred == "a == b" &&
+                   isArray!R &&
+                   allSatisfy!(isArray, Ss) &&
+                   allSatisfy!(isEquableToR, Ss))
         {
-            return r1 == r2;
+            foreach (ref s; ss)
+            {
+                if (r != s) return false;
+            }
+            return true;
         }
-        //Try a fast implementation when the ranges have comparable lengths
-        else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+        // use fast implementation when the ranges have comparable lengths
+        else static if (allSatisfy!(hasLength, Ss) &&
+                        allSatisfy!(hasLengthComparableToR, Ss))
         {
-            auto len1 = r1.length;
-            auto len2 = r2.length;
-            if (len1 != len2) return false; //Short circuit return
-
-            //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can sqeeze out a bit of performance by not checking if r2 is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+            // check equal lengths
+            foreach (ref s; ss)
             {
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                if (r.length != s.length) { return false; }
+            }
+            // check equal contents
+            foreach (ref e; r)
```

Comment:
Here is the bug - for strings it would iterate by char/wchar, so keep the original ugly `for`.